velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -75,53 +75,64 @@ void PageReader::seekToPage(int64_t row) {
   }
 }
 
+// The size of Parquet Page Header remains unknown until the Header is parsed
+// because it is in Thrift format. To read Parquet Page Header, we are using an
+// adaptive approach to guess the Page Header size.
+
+// TODO: Refactor the inputStreams to have a general interface where read() can
+// be applied on different chunks of memory.
 PageHeader PageReader::readPageHeader(int64_t remainingSize) {
   // Note that sizeof(PageHeader) may be longer than actually read
   std::shared_ptr<thrift::ThriftBufferedTransport> transport;
   std::unique_ptr<apache::thrift::protocol::TCompactProtocolT<
       thrift::ThriftBufferedTransport>>
       protocol;
-  char copy[sizeof(PageHeader)];
-  bool wasInBuffer = false;
-  if (bufferEnd_ == bufferStart_) {
-    const void* buffer;
-    int32_t size;
-    inputStream_->Next(&buffer, &size);
-    bufferStart_ = reinterpret_cast<const char*>(buffer);
-    bufferEnd_ = bufferStart_ + size;
-  }
-  if (bufferEnd_ - bufferStart_ >= sizeof(PageHeader)) {
-    wasInBuffer = true;
-    transport = std::make_shared<thrift::ThriftBufferedTransport>(
-        bufferStart_, sizeof(PageHeader));
-    protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
-        thrift::ThriftBufferedTransport>>(transport);
-  } else {
+
+  uint32_t pageHeaderBufferSize = kPageHeaderBufferSize_;
+  uint32_t oldPageHeaderBufferSize = 0;
+
+  BufferPtr copyBuffer =
+      AlignedBuffer::allocate<char>(pageHeaderBufferSize, &pool_);
+
+  uint32_t attempt = 0;
+  for (attempt = 0; attempt < kMaxAttempts_; attempt++) {
     dwio::common::readBytes(
-        std::min<int64_t>(remainingSize, sizeof(PageHeader)),
+        std::min<int64_t>(
+            remainingSize, pageHeaderBufferSize - oldPageHeaderBufferSize),
         inputStream_.get(),
-        &copy,
+        copyBuffer->asMutable<char>() + oldPageHeaderBufferSize,
         bufferStart_,
         bufferEnd_);
-
     transport = std::make_shared<thrift::ThriftBufferedTransport>(
-        copy, sizeof(PageHeader));
+        copyBuffer->asMutable<char>(), pageHeaderBufferSize);
     protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
         thrift::ThriftBufferedTransport>>(transport);
+
+    try {
+      PageHeader pageHeader;
+      // Use the Thrift generated read() function to read Parquet page header.
+      uint64_t readBytes = pageHeader.read(protocol.get());
```

Comment:
Will this read and parse the PageHeader multiple times if it's larger than 400 bytes? Parsing the PageHeader is already expensive and this will cause additional performance regressions. A more proper way is to change the ThriftBufferedTransport such that it expands the buffer when the requested size (the len in `uint32_t read(uint8_t* outputBuf, uint32_t len)`) is greater than the current buffer size.

Reply:
This retry should be happening in rare cases so it's probably fine.  If we are going to change `ThriftBufferedTransport`, it's rather better to enhance `TCompactProtocolT` to tell us the exact header size so that we do not have to do this retry.

Reply:
Yes. The first attempt is 512 bytes.  In our analysis, most of the headers are smaller that 512. The header size becomes large only when the column data is string and the corresponding min and max are pretty large.

And we can know the header size only when it is fully parsed, because it is in Thrift format.